### PR TITLE
Pie Chart

### PR DESCRIPTION
**Pull request by vkravchik**

### CODE REVIEW

File: server/routes/pieRoutes.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  let json = require('../dataset/mapped_pie_data');
+
+  const start_date = req.query.start_date || 1999;
+  const finish_date = req.query.finish_date || 2010;
+
+  // TODO: Condition for range < 10 years
+
+  const data = [];
+
+  _.forEach(json, (el, year) => {
```
Comment (ivan629) [inline server/routes/pieRoutes.js]: please el => element

File: server/routes/utilRoutes.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 const router = require('express').Router();
-const {parseCsvToJson, remapCurrentJson} = require('../utils/parseCsvToJson');
+const {parseCsvToJson, remapCurrentJson, mapPieJson} = require('../utils/parseCsvToJson');
```
Comment (ivan629) [inline server/routes/utilRoutes.js]: please fix formatting

File: server/utils/parseCsvToJson.js [hunk 3]
```diff
@@ -1,42 +1,55 @@
 const csvToJson = require('convert-csv-to-json');
-const router = require('express').Router();
 const path = require('path');
 const fs = require('fs');
```
Comment (ivan629) [inline server/utils/parseCsvToJson.js:1]: please move `fs` module to the first line, as it is a native module

File: src/App.js [hunk 4]
```diff
@@ -4,6 +4,7 @@ import 'bootstrap/dist/css/bootstrap.min.css';
 import 'antd/dist/antd.css';
 
 import Chart from './containers/Chart';
+import Pie from './containers/Pie';
```
Comment (ivan629) [inline src/App.js]: please rename `Pie => PieChart`, `Chart => StockChart`

File: src/common/constants/pieConstants.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export const GET_PIE_REQUEST = 'GET_PIE_REQUEST';
```
Comment (ivan629) [inline src/common/constants/pieConstants.js]: please rename `GET_PIE_REQUEST => PIE_REQUEST_GET` and al other constants

File: src/common/constants/highStockConstants.js [hunk 6]
```diff
@@ -2,6 +2,7 @@ export const GET_HIGH_STOCK_REQUEST = 'GET_HIGH_STOCK_REQUEST';
 export const GET_HIGH_STOCK_REQUEST_SUCCESS = 'GET_HIGH_STOCK_REQUEST_SUCCESS';
```
Comment (ivan629) [inline src/common/constants/highStockConstants.js]: why do you need this const ? consider to get rid of `GET_HIGH_STOCK_REQUEST_SUCCESS `

File: src/components/Error.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+export const Error = ({error}) => {
+  return (
+    <>
```
Comment (ivan629) [inline src/components/Error.js]: please delete `<>`

File: src/components/PieChart.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import { getPieAction } from '../redux/actions/pieActions';
+import { Error } from "./Error";
+
+const PieChart = (props) => {
+  const {getPieAction} = props;
+  const {isLoading, data, error} = props.pieProps;
+  const {initialConfig: {defaultStart, defaultFinish}} = props.sliderProps;
+
+
+  useEffect(() => {
```
Comment (ivan629) [inline src/components/PieChart.js:17]: one line please

File: src/components/PieChart.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      }
+    },
+    series: [{
+      name: 'Observations',
+      colorByPoint: true,
+      data: data,
+    }]
+  };
+
+  return (
```
Comment (ivan629) [inline src/components/PieChart.js:59]: `return error ? <Error error={error} />  : <ReactHighcharts config={config} />`

File: src/App.js (hunk 4 above)
Comment (vkravchik) [reply src/App.js]: Sorry but I have components with names `PieChart` and `HighStockChart`. Could u please say how I must name components and containers?

File: src/containers/Chart.js [hunk 10]
```diff
@@ -3,23 +3,34 @@ import React, { useEffect } from 'react';
...
 
 const Chart = (props) => {
-
-  const {getHighStockConfigAction} = props;
-  const {data, error, initialConfig} = props.highStockReducer;
+  const {getSliderConfigAction, getHighStockAction, setSliderConfigAction} = props;
+  const {data, error} = props.highStockProps;
+  const {initialConfig} = props.sliderProps;
 
   useEffect(() => {
```
Comment (ivan629) [inline src/containers/Chart.js]: one line please

File: src/containers/Chart.js [hunk 11]
```diff
@@ -3,23 +3,34 @@ import React, { useEffect } from 'react';
...
-    getHighStockConfigAction();
-  }, [getHighStockConfigAction]);
+    getSliderConfigAction();
+  }, [getSliderConfigAction]);
+
+  const onAfterChange = (value = []) => {
+    const dateObj = {};
```
Comment (ivan629) [inline src/containers/Chart.js]: dateObj => dateMap

File: src/containers/Chart.js [hunk 12]
```diff
@@ -40,14 +51,13 @@ const Chart = (props) => {
...
   return {
-    highStockReducer: state.highStockReducer,
-  }
-};
-
-const mapDispatchToProps = (dispatch) => {
-  return {
-    getHighStockConfigAction: () => dispatch(getHighStockConfigAction()),
+    highStockProps: state.highStockReducer,
+    sliderProps: state.sliderReducer,
```
Comment (ivan629) [inline src/containers/Chart.js]: please delete comma in the end

File: src/containers/Pie.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import { connect } from 'react-redux';
+import PieChart from '../components/PieChart';
+import RangeSlider from '../components/RangeSlider';
+
+import { getSliderConfigAction, setSliderConfigAction } from '../redux/actions/sliderActions';
+import { getPieAction } from '../redux/actions/pieActions';
+
+const Pie = (props) => {
+  const {getSliderConfigAction, getPieAction, setSliderConfigAction} = props;
+  const {initialConfig} = props;
```
Comment (ivan629) [inline src/containers/Pie.js]: fix formatting

File: src/containers/Pie.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+  useEffect(() => {
```
Comment (ivan629) [inline src/containers/Pie.js]: one line please

File: src/containers/Pie.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    getSliderConfigAction();
+  },[getSliderConfigAction]);
+
+  const onAfterChange = (value = []) => {
+    const dateObj = {};
```
Comment (ivan629) [inline src/containers/Pie.js]: dateObj => dateMap

File: src/containers/Pie.js [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+    dateObj['start_date'] = value[0];
+    dateObj['finish_date'] = value[1];
+
+    setSliderConfigAction(dateObj);
+    getPieAction(dateObj);
+  };
+
+  return (
+    <>
```
Comment (ivan629) [inline src/containers/Pie.js]: delete unnecessary `<>`

File: src/redux/reducers/index.js [hunk 17]
```diff
@@ -1,9 +1,13 @@
...
 import { highStockReducer } from './highStockReducer';
-import { switchTypeReducer } from "./switchTypeReducer";
+import { switchTypeReducer } from './switchTypeReducer';
+import { sliderReducer } from './sliderReducer';
 
 export default combineReducers({
   highStockReducer: highStockReducer,
+  pieReducer: pieReducer,
   switchTypeReducer: switchTypeReducer,
+  sliderReducer: sliderReducer,
```
Comment (ivan629) [inline src/redux/reducers/index.js]: extra comma in the end

File: src/redux/reducers/pieReducer.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
+import { GET_PIE_REQUEST, GET_PIE_REQUEST_ERROR, GET_PIE_REQUEST_SUCCESS } from '../../common/constants/pieConstants';
+
+const initialState = {
+  isLoading: false,
+  data: [],
+  error: null,
```
Comment (ivan629) [inline src/redux/reducers/pieReducer.js]: extra comma in the end

File: src/redux/reducers/pieReducer.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+export const pieReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case GET_PIE_REQUEST: {
+      return {
+        ...state,
+        isLoading: true,
```
Comment (ivan629) [inline src/redux/reducers/pieReducer.js]: delete comma

File: src/redux/reducers/sliderReducer.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+const initialState = {
+  isLoading: false,
+  initialConfig: {
+    min: null,
+    max: null,
+    defaultStart: null,
+    defaultFinish: null
+  },
+  error: null,
```
Comment (ivan629) [inline src/redux/reducers/sliderReducer.js]: delete extra comma

File: src/redux/reducers/sliderReducer.js [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+        ...state,
+        isLoading: true
+      };
+
+    case SET_HIGH_STOCK_CONFIG_REQUEST:
+      return {
+        ...state,
+        initialConfig: {
+          ...state.initialConfig,
+          defaultStart: action.payload?.start_date || null,
```
Comment (ivan629) [inline src/redux/reducers/sliderReducer.js:31]: what is this `action.payload?.start_date`?
use `isNil` from the `lodash` library and ternary operator

File: src/redux/sagas/switchTypeSaga/switchTypeSagaWorker.js [hunk 22]
```diff
@@ -1,5 +1,5 @@
-import { put } from "redux-saga//effects";
-import { TOGGLE_CHART } from "../../../common/constants/switchTypeConstants";
+import { put } from 'redux-saga/effects';
+import { TOGGLE_CHART } from '../../../common/constants/switchTypeConstants';
 
 export function* toggleChartTypeWorker({payload}) {
   yield put({type: TOGGLE_CHART, payload})
```
Comment (ivan629) [inline src/redux/sagas/switchTypeSaga/switchTypeSagaWorker.js]: fix formatting

File: src/App.js (hunk 4 above)
Comment (ivan629) [reply src/App.js]: containers: PieChartContainer,
components: PieChart

File: src/components/PieChart.js [hunk 23]
```diff
@@ -0,0 +1,75 @@
+import React, { useEffect } from 'react';
+import { connect } from 'react-redux';
+import ReactHighcharts from 'react-highcharts/';
+
+import { getPieAction } from '../redux/actions/pieActions';
+import { Error } from "./Error";
```
Comment (ivan629) [inline src/components/PieChart.js]: use singlequote

File: src/components/PieChart.js [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+  useEffect(() => {
+    getPieAction();
+  }, [getPieAction]);
+
+  const config = {
+    chart: {
+      events: {
+        load: function () {
+          this.showLoading();
```
Comment (ivan629) [inline src/components/PieChart.js:25]: add new after

File: src/components/PieChart.js [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+        dataLabels: {
+          enabled: false
+        },
+        showInLegend: true
+      }
+    },
+    series: [{
+      name: 'Observations',
+      colorByPoint: true,
+      data: data,
```
Comment (ivan629) [inline src/components/PieChart.js:55]: delete extra comma

**Review by ivan629 — CHANGES_REQUESTED** [review]
fix it

File: src/components/PieChart.js (hunk 24 above)
Comment (vkravchik) [reply src/components/PieChart.js:25]: what did u mean?

File: src/common/constants/pieConstants.js (hunk 5 above)
Comment (vkravchik) [reply src/common/constants/pieConstants.js]: I think it's useless to do that. I can spend time more useful for myself. In this case, it's not much necessary.
For example, I can spend time for below comment

File: src/App.js [hunk 26]
```diff
@@ -3,8 +3,9 @@ import { BrowserRouter as Router, Route } from 'react-router-dom';
 import 'bootstrap/dist/css/bootstrap.min.css';
```
Comment (ivan629) [inline src/App.js:3]: please delete `bootstrap`

File: src/components/RangeSlider.js [hunk 27]
```diff
@@ -30,7 +19,8 @@ const RangeSlider = (props) => {
                   onAfterChange={onAfterChange}/>
         </div>
         <div className="col-md-1 text-center filter-container">
-          <Button className='align-icon' type="primary" shape="round" icon="filter" size='small' onClick={onApplyFilter}/>
+          <Button className='align-icon' type="primary" shape="round" icon="filter" size='small'
```
Comment (ivan629) [inline src/components/RangeSlider.js]: format it

File: src/components/SwitchType.js [hunk 28]
```diff
@@ -31,8 +31,6 @@ const mapStateToProps = (state, ownProps) => ({
...
 });
 
-const mapDispatchToProps = (dispatch) => ({
-  toggleChartTypeAction: (type) => dispatch(toggleChartTypeAction(type)),
-});
-
-export default connect(mapStateToProps, mapDispatchToProps)(SwitchType)
+export default connect(mapStateToProps, {
+  toggleChartTypeStart
+})(SwitchType)
```
Comment (ivan629) [inline src/components/SwitchType.js:36]: ;

File: src/containers/HighStockChartContainer.js [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+import '../common/styles/Chart.scss';
+
+import { getHighStockData } from '../redux/actions/highStockActions';
+import { getSliderConfigData, setSliderConfigData } from '../redux/actions/sliderActions';
+
+import HighStockChart from '../components/HighStockChart';
+import RangeSlider from '../components/RangeSlider';
+
+const HighStockChartContainer = (props) => {
+  const {
```
Comment (ivan629) [inline src/containers/HighStockChartContainer.js:13]: reformat

File: src/redux/reducers/sliderReducer.js [hunk 30]
```diff
@@ -0,0 +1,57 @@
+import {
+  HIGH_STOCK_CONFIG_FETCH, HIGH_STOCK_CONFIG_FETCH_ERROR,
+  HIGH_STOCK_CONFIG_FETCH_SUCCESS, HIGH_STOCK_CONFIG_SET
+} from '../../common/constants/highStockConstants';
+
+const initialState = {
+  isLoading: false,
```
Comment (ivan629) [inline src/redux/reducers/sliderReducer.js:8]:
```
 error: null,
 isLoading: false,
 initialConfig: {
    min: null,
    max: null,
    defaultStart: null,
    defaultFinish: null
  }
```

File: src/redux/reducers/switchTypeReducer.js [hunk 31]
```diff
@@ -1,17 +1,17 @@
-import { TOGGLE_CHART } from "../../common/constants/switchTypeConstants";
+import { TOGGLE_CHART_TYPE_FINISH } from '../../common/constants/switchTypeConstants';
 
 const initialState = {
-  highStockType: 'area',
+  highStockType: 'area'
```
Comment (ivan629) [inline src/redux/reducers/switchTypeReducer.js]: `highStockType => highStockChartType`

**Review by ivan629 — CHANGES_REQUESTED** [review]
fix it

File: src/App.js (hunk 26 above)
Comment (vkravchik) [reply src/App.js:3]: Sorry, I will remove this later a little bit. At the moment I need it for some classes

**Pull request merged**